flax/nnx/nn/normalization.py

```diff
@@ -988,3 +1004,200 @@ def __call__(self, x, *, mask: tp.Optional[jax.Array] = None):
       self.dtype,
       self.epsilon,
     )
+
+
+class SpectralNorm(Module):
+  """Spectral normalization.
+
+  See:
+
+  - https://arxiv.org/abs/1802.05957
+  - https://arxiv.org/abs/1805.08318
+  - https://arxiv.org/abs/1809.11096
+
+  Spectral normalization normalizes the weight params so that the spectral
+  norm of the matrix is equal to 1. This is implemented as a layer wrapper
+  where each wrapped layer will have its params spectral normalized before
+  computing its ``__call__`` output.
+
+  .. note::
+    The initialized variables dict will contain, in addition to a 'params'
+    collection, a separate 'batch_stats' collection that will contain a
+    ``u`` vector and ``sigma`` value, which are intermediate values used
+    when performing spectral normalization. During training, we pass in
+    ``update_stats=True`` so that ``u`` and ``sigma`` are updated with
+    the most recently computed values using power iteration. This will
+    help the power iteration method approximate the true singular value
+    more accurately over time. During eval, we pass in ``update_stats=False``
+    to ensure we get deterministic behavior from the model.
+
+  Example usage::
+
+    >>> from flax import nnx
+    >>> import jax
+    >>> rngs = nnx.Rngs(0)
+    >>> x = jax.random.normal(jax.random.key(0), (3, 4))
+    >>> layer = nnx.SpectralNorm(nnx.Linear(4, 5, rngs=rngs), rngs=rngs)
+    >>> jax.tree.map(jax.numpy.shape, nnx.state(layer, nnx.Param))
+    State({
+      'layer_instance': {
+        'bias': Param(
+          value=(5,)
+        ),
+        'kernel': Param(
+          value=(4, 5)
+        )
+      }
+    })
+    >>> y = layer(x, update_stats=True)
+
+  Args:
+    layer_instance: Module instance that is wrapped with SpectralNorm
+    n_steps: How many steps of power iteration to perform to approximate the
+      singular value of the weight params.
+    epsilon: A small float added to l2-normalization to avoid dividing by zero.
+    dtype: the dtype of the result (default: infer from input and params).
+    param_dtype: the dtype passed to parameter initializers (default: float32).
+    error_on_non_matrix: Spectral normalization is only defined on matrices. By
+      default, this module will return scalars unchanged and flatten
+      higher-order tensors in their leading dimensions. Setting this flag to
+      True will instead throw an error if a weight tensor with dimension greater
+      than 2 is used by the layer.
+    update_stats: if True, the stored batch statistics will be
+      used instead of computing the batch statistics on the input.
+    rngs: rng key.
+  """
+
+  def __init__(
+    self,
+    layer_instance: Module,
+    *,
+    n_steps: int = 1,
+    epsilon: float = 1e-12,
+    dtype: tp.Optional[Dtype] = None,
+    param_dtype: Dtype = jnp.float32,
+    error_on_non_matrix: bool = False,
+    update_stats: bool = True,
+    rngs: rnglib.Rngs,
+  ):
+    self.layer_instance = layer_instance
```

Comment:
Should we check for the type of `layer_instance`? 
It should be a subclass of `nnx.Module` to have parameters and be a callable.